File: native-android/app/src/main/java/fr/eiter/plexiwine/RpgModels.kt
package fr.eiter.plexiwine

import com.google.gson.annotations.SerializedName
import com.google.gson.JsonElement

/** GET /api/rpg/me */
data class RpgState(
    val enabled: Boolean = false,
    val ui: Boolean = false,
    val allowed: Boolean = false,
    val profile: RpgProfile? = null,
    val quests: RpgQuests? = null,
    val badges: List<RpgBadge> = emptyList(),
    @SerializedName("next_badges") val nextBadges: List<RpgBadge> = emptyList(),
    val atlas: RpgAtlas? = null,
    val classes: List<RpgClassInfo> = emptyList(),
    @SerializedName("class_affinity") val classAffinity: Map<String, Int>? = null,
    val phrase: String? = null
)

data class RpgProfile(
    val username: String? = null,
    val level: Int = 1,
    val xp: Int = 0,
    val title: String? = null,
    @SerializedName("progress_pct") val progressPct: Double = 0.0,
    @SerializedName("xp_to_next") val xpToNext: Int? = null,
    @SerializedName("xp_into_level") val xpIntoLevel: Int? = null,
    @SerializedName("xp_level_start") val xpLevelStart: Int? = null,
    @SerializedName("xp_level_next") val xpLevelNext: Int? = null,
    @SerializedName("streak_days") val streakDays: Int = 0,
    @SerializedName("daily_xp") val dailyXp: Int = 0,
    @SerializedName("daily_soft_cap") val dailySoftCap: Int = 100,
    @SerializedName("daily_soft_capped") val dailySoftCapped: Boolean = false,
    @SerializedName("daily_soft_cap_remaining") val dailySoftCapRemaining: Int? = null,
    @SerializedName("class") val classKey: String? = null,
    @SerializedName("class_info") val classInfo: RpgClassInfo? = null,
    @SerializedName("beer_master") val beerMaster: Boolean = false,
    val prestige: RpgPrestige? = null,
    @SerializedName("title_band") val titleBand: RpgTitleBand? = null,
    @SerializedName("intro_seen") val introSeen: Boolean = false
)

data class RpgClassInfo(
    val key: String? = null,
    val name: String? = null,
    val icon: String? = null,
    val blurb: String? = null,
    @SerializedName("when") val whenText: String? = null,
    val special: String? = null
)

data class RpgPrestige(
    val key: String? = null,
    val icon: String? = null,
    val ribbon: String? = null,
    val tagline: String? = null,
    val blurb: String? = null
)

data class RpgTitleBand(
    val name: String? = null,
    @SerializedName("from") val fromLevel: Int? = null,
    val to: Int? = null
)

data class RpgQuests(
    val active: List<RpgQuest> = emptyList(),
    @SerializedName("done_today") val doneToday: List<RpgQuest> = emptyList(),
    @SerializedName("done_weekly") val doneWeekly: List<RpgQuest> = emptyList()
)

data class RpgQuest(
    val key: String? = null,
    val kind: String? = null,
    val title: String? = null,
    val description: String? = null,
    val progress: Int = 0,
    val target: Int = 1,
    val status: String? = null,
    @SerializedName("reward_xp") val rewardXp: Int = 0
)

data class RpgBadge(
    val key: String? = null,
    val name: String? = null,
    val icon: String? = null,
    val rarity: String? = null,
    val lore: String? = null,
    val hint: String? = null,
    val earned: Boolean = false,
    @SerializedName("earned_at") val earnedAt: String? = null,
    val progress: Int = 0,
    val target: Int = 1,
    val remaining: Int? = null,
    @SerializedName("unlock_phrase") val unlockPhrase: String? = null
)

data class RpgAtlas(
    @SerializedName("styles_count") val stylesCount: Int = 0,
    @SerializedName("hops_count") val hopsCount: Int = 0,
    @SerializedName("breweries_count") val breweriesCount: Int = 0,
    val photos: Int = 0,
    @SerializedName("total_checkins") val totalCheckins: Int = 0,
    val styles: List<String>? = null
)

/** Bloc `rpg` renvoyé par POST /api/checkins */
data class RpgLoot(
    @SerializedName("xp_gained") val xpGained: Int = 0,
    val xp: Int = 0,
    val level: Int = 1,
    @SerializedName("level_up") val levelUp: Boolean = false,
    @SerializedName("old_level") val oldLevel: Int? = null,
    @SerializedName("levels_gained") val levelsGained: Int? = null,
    val title: String? = null,
    @SerializedName("old_title") val oldTitle: String? = null,
    @SerializedName("title_changed") val titleChanged: Boolean = false,
    @SerializedName("progress_pct") val progressPct: Double = 0.0,
    @SerializedName("xp_to_next") val xpToNext: Int? = null,
    val phrase: String? = null,
    @SerializedName("phrase_level_up") val phraseLevelUp: String? = null,
    @SerializedName("badges_earned") val badgesEarned: List<RpgBadge> = emptyList(),
    @SerializedName("quests_completed") val questsCompleted: List<RpgQuest> = emptyList(),
    @SerializedName("next_badges") val nextBadges: List<RpgBadge> = emptyList(),
    @SerializedName("streak_days") val streakDays: Int? = null,
    @SerializedName("daily_xp") val dailyXp: Int? = null,
    @SerializedName("daily_soft_cap") val dailySoftCap: Int? = null,
    @SerializedName("daily_soft_capped") val dailySoftCapped: Boolean = false,
    @SerializedName("daily_soft_cap_just_hit") val dailySoftCapJustHit: Boolean = false,
    @SerializedName("soft_cap_message") val softCapMessage: String? = null,
    /** breakdown items may be heterogeneous — keep as JsonElement list if needed */
    val breakdown: List<Map<String, JsonElement>>? = null
)

/** File de célébrations (level-up / badge). */
sealed class RpgCelebration {
    data class LevelUp(val loot: RpgLoot) : RpgCelebration()
    data class BadgeUnlock(val badge: RpgBadge) : RpgCelebration()
}

data class RpgAdminPlayersResponse(
    val players: List<RpgAdminPlayer> = emptyList(),
    val total: Int = 0,
    @SerializedName("with_profile") val withProfile: Int = 0,
    val flags: RpgAdminFlags? = null,
)

data class RpgAdminFlags(
    val enabled: Boolean = false,
    val ui: Boolean = false,
    val backfill: Boolean = false,
    @SerializedName("allow_invites") val allowInvites: Boolean = false,
    val allowlist: List<String> = emptyList(),
    val mutable: Boolean = true,
)

data class RpgAdminSettingsResponse(
    val ok: Boolean = true,
    val flags: RpgAdminFlags? = null,
)

data class RpgAdminPlayer(
    val username: String? = null,
    val level: Int = 1,
    val xp: Int = 0,
    val title: String? = null,
    @SerializedName("streak_days") val streakDays: Int = 0,
    @SerializedName("class") val classKey: String? = null,
    @SerializedName("beer_master") val beerMaster: Boolean = false,
    @SerializedName("is_admin") val isAdmin: Boolean = false,
    @SerializedName("is_invite") val isInvite: Boolean = false,
    val checkins: Int = 0,
    @SerializedName("badge_count") val badgeCount: Int = 0,
    val allowed: Boolean = true,
    /** true/false = override admin, null = règles défaut (allowlist/env). Gson: missing → null. */
    @SerializedName("allowed_override") val allowedOverride: Boolean? = null,
    @SerializedName("has_profile") val hasProfile: Boolean = false,
    @SerializedName("daily_soft_cap") val dailySoftCap: Int = 0,
    @SerializedName("daily_xp_today") val dailyXpToday: Int = 0,
    @SerializedName("daily_checkins_today") val dailyCheckinsToday: Int = 0,
    @SerializedName("daily_soft_capped") val dailySoftCapped: Boolean = false,
    @SerializedName("daily_soft_cap_remaining") val dailySoftCapRemaining: Int? = null
)

fun RpgProfile.displayIcon(): String {
    if (beerMaster) return prestige?.icon ?: "👑"
    return classInfo?.icon ?: "🍷"
}

fun rarityLabelFr(r: String?): String = when ((r ?: "common").lowercase()) {
    "legendary" -> "Légendaire"
    "epic" -> "Épique"
    "rare" -> "Rare"
    else -> "Commun"
}

File: native-ios/WineNative/Sources/RpgModels.swift
import Foundation

struct RpgState: Decodable {
    var enabled: Bool?
    var ui: Bool?
    var allowed: Bool?
    var profile: RpgProfile?
    var quests: RpgQuests?
    var badges: [RpgBadge]?
    var nextBadges: [RpgBadge]?
    var atlas: RpgAtlas?
    var classes: [RpgClassInfo]?
    var classAffinity: [String: Int]?
    var phrase: String?

    enum CodingKeys: String, CodingKey {
        case enabled, ui, allowed, profile, quests, badges, atlas, classes, phrase
        case nextBadges = "next_badges"
        case classAffinity = "class_affinity"
    }

    var active: Bool {
        enabled == true && ui == true && profile != nil
    }
}

struct RpgProfile: Decodable {
    var username: String?
    var level: Int?
    var xp: Int?
    var title: String?
    var progressPct: Double?
    var xpToNext: Int?
    var xpIntoLevel: Int?
    var xpLevelStart: Int?
    var xpLevelNext: Int?
    var streakDays: Int?
    var dailyXp: Int?
    var dailySoftCap: Int?
    var dailySoftCapped: Bool?
    var dailySoftCapRemaining: Int?
    var classKey: String?
    var classInfo: RpgClassInfo?
    var beerMaster: Bool?
    var prestige: RpgPrestige?
    var titleBand: RpgTitleBand?
    var introSeen: Bool?

    enum CodingKeys: String, CodingKey {
        case username, level, xp, title, prestige
        case progressPct = "progress_pct"
        case xpToNext = "xp_to_next"
        case xpIntoLevel = "xp_into_level"
        case xpLevelStart = "xp_level_start"
        case xpLevelNext = "xp_level_next"
        case streakDays = "streak_days"
        case dailyXp = "daily_xp"
        case dailySoftCap = "daily_soft_cap"
        case dailySoftCapped = "daily_soft_capped"
        case dailySoftCapRemaining = "daily_soft_cap_remaining"
        case classKey = "class"
        case classInfo = "class_info"
        case beerMaster = "beer_master"
        case titleBand = "title_band"
        case introSeen = "intro_seen"
    }

    var displayIcon: String {
        if beerMaster == true { return prestige?.icon ?? "👑" }
        return classInfo?.icon ?? "🍷"
    }
}

struct RpgClassInfo: Decodable, Identifiable {
    var key: String?
    var name: String?
    var icon: String?
    var blurb: String?
    var whenText: String?
    var special: String?
    var id: String { key ?? name ?? UUID().uuidString }

    enum CodingKeys: String, CodingKey {
        case key, name, icon, blurb, special
        case whenText = "when"
    }
}

struct RpgPrestige: Decodable {
    var key: String?
    var icon: String?
    var ribbon: String?
    var tagline: String?
    var blurb: String?
}

struct RpgTitleBand: Decodable {
    var name: String?
    var fromLevel: Int?
    var to: Int?
    enum CodingKeys: String, CodingKey {
        case name, to
        case fromLevel = "from"
    }
}

struct RpgQuests: Decodable {
    var active: [RpgQuest]?
    var doneToday: [RpgQuest]?
    var doneWeekly: [RpgQuest]?
    enum CodingKeys: String, CodingKey {
        case active
        case doneToday = "done_today"
        case doneWeekly = "done_weekly"
    }
}

struct RpgQuest: Decodable, Identifiable {
    var key: String?
    var kind: String?
    var title: String?
    var description: String?
    var progress: Int?
    var target: Int?
    var status: String?
    var rewardXp: Int?
    var id: String { key ?? title ?? UUID().uuidString }
    enum CodingKeys: String, CodingKey {
        case key, kind, title, description, progress, target, status
        case rewardXp = "reward_xp"
    }
}

struct RpgBadge: Decodable, Identifiable {
    var key: String?
    var name: String?
    var icon: String?
    var rarity: String?
    var lore: String?
    var hint: String?
    var earned: Bool?
    var earnedAt: String?
    var progress: Int?
    var target: Int?
    var remaining: Int?
    var unlockPhrase: String?
    var id: String { key ?? name ?? UUID().uuidString }
    enum CodingKeys: String, CodingKey {
        case key, name, icon, rarity, lore, hint, earned, progress, target, remaining
        case earnedAt = "earned_at"
        case unlockPhrase = "unlock_phrase"
    }
}

struct RpgAtlas: Decodable {
    var stylesCount: Int?
    var hopsCount: Int?
    var breweriesCount: Int?
    var photos: Int?
    var totalCheckins: Int?
    var styles: [String]?
    enum CodingKeys: String, CodingKey {
        case photos, styles
        case stylesCount = "styles_count"
        case hopsCount = "hops_count"
        case breweriesCount = "breweries_count"
        case totalCheckins = "total_checkins"
    }
}

struct RpgLoot: Decodable {
    var xpGained: Int?
    var xp: Int?
    var level: Int?
    var levelUp: Bool?
    var oldLevel: Int?
    var levelsGained: Int?
    var title: String?
    var oldTitle: String?
    var titleChanged: Bool?
    var progressPct: Double?
    var xpToNext: Int?
    var phrase: String?
    var phraseLevelUp: String?
    var badgesEarned: [RpgBadge]?
    var questsCompleted: [RpgQuest]?
    var nextBadges: [RpgBadge]?
    var streakDays: Int?
    var dailyXp: Int?
    var dailySoftCap: Int?
    var dailySoftCapped: Bool?
    var dailySoftCapJustHit: Bool?
    var softCapMessage: String?
    enum CodingKeys: String, CodingKey {
        case xp, level, title, phrase
        case xpGained = "xp_gained"
        case levelUp = "level_up"
        case oldLevel = "old_level"
        case levelsGained = "levels_gained"
        case oldTitle = "old_title"
        case titleChanged = "title_changed"
        case progressPct = "progress_pct"
        case xpToNext = "xp_to_next"
        case phraseLevelUp = "phrase_level_up"
        case badgesEarned = "badges_earned"
        case questsCompleted = "quests_completed"
        case nextBadges = "next_badges"
        case streakDays = "streak_days"
        case dailyXp = "daily_xp"
        case dailySoftCap = "daily_soft_cap"
        case dailySoftCapped = "daily_soft_capped"
        case dailySoftCapJustHit = "daily_soft_cap_just_hit"
        case softCapMessage = "soft_cap_message"
    }
}

// MARK: - Célébrations / intro

enum RpgCelebration: Identifiable, Equatable {
    case levelUp(RpgLoot)
    case badge(RpgBadge)

    var id: String {
        switch self {
        case .levelUp(let loot):
            return "lvl-\(loot.level ?? 0)-\(loot.oldLevel ?? 0)-\(loot.xp ?? 0)"
        case .badge(let b):
            return "badge-\(b.key ?? b.name ?? UUID().uuidString)"
        }
    }

    static func == (lhs: RpgCelebration, rhs: RpgCelebration) -> Bool {
        lhs.id == rhs.id
    }
}

// MARK: - Admin Weeno Quest (liste joueurs)

struct RpgAdminPlayersResponse: Decodable {
    var players: [RpgAdminPlayer]?
    var total: Int?
    var withProfile: Int?
    var flags: RpgAdminFlags?
    enum CodingKeys: String, CodingKey {
        case players, total, flags
        case withProfile = "with_profile"
    }
}

struct RpgAdminFlags: Decodable, Equatable {
    var enabled: Bool?
    var ui: Bool?
    var backfill: Bool?
    var allowInvites: Bool?
    var allowlist: [String]?
    var mutable: Bool?

    enum CodingKeys: String, CodingKey {
        case enabled, ui, backfill, allowlist, mutable
        case allowInvites = "allow_invites"
    }
}

struct RpgAdminSettingsResponse: Decodable {
    var ok: Bool?
    var flags: RpgAdminFlags?
}

struct RpgAdminPlayer: Decodable, Identifiable {
    var username: String?
    var level: Int?
    var xp: Int?
    var title: String?
    var streakDays: Int?
    var classKey: String?
    var classInfo: RpgClassInfo?
    var beerMaster: Bool?
    var isAdmin: Bool?
    var isInvite: Bool?
    var checkins: Int?
    var badgeCount: Int?
    var allowed: Bool?
    /// true/false = override admin ; nil = règles défaut (allowlist/env)
    var allowedOverride: Bool?
    var hasProfile: Bool?
    var progressPct: Double?
    var suspicionScore: Int?
    var suspicionFlagged: Bool?
    var orphan: Bool?
    var introSeen: Bool?
    var backfilled: Bool?
    var dailyXpTotal: Int?
    var dailyXpCount: Int?
    var dailySoftCap: Int?
    var dailyXpToday: Int?
    var dailyCheckinsToday: Int?
    var dailySoftCapped: Bool?
    var dailySoftCapRemaining: Int?
    var lastRpgCheckinAt: String?
    var id: String { username ?? UUID().uuidString }

    enum CodingKeys: String, CodingKey {
        case username, level, xp, title, checkins, allowed, orphan
        case allowedOverride = "allowed_override"
        case streakDays = "streak_days"
        case classKey = "class"
        case classInfo = "class_info"
        case beerMaster = "beer_master"
        case isAdmin = "is_admin"
        case isInvite = "is_invite"
        case badgeCount = "badge_count"
        case hasProfile = "has_profile"
        case progressPct = "progress_pct"
        case suspicionScore = "suspicion_score"
        case suspicionFlagged = "suspicion_flagged"
        case introSeen = "intro_seen"
        case backfilled
        case dailyXpTotal = "daily_xp_total"
        case dailyXpCount = "daily_xp_count"
        case dailySoftCap = "daily_soft_cap"
        case dailyXpToday = "daily_xp_today"
        case dailyCheckinsToday = "daily_checkins_today"
        case dailySoftCapped = "daily_soft_capped"
        case dailySoftCapRemaining = "daily_soft_cap_remaining"
        case lastRpgCheckinAt = "last_rpg_checkin_at"
    }
}

/// GET/PATCH /api/admin/rpg/players/{user} — détail complet
struct RpgAdminPlayerDetail: Decodable {
    var player: RpgAdminPlayer?
    var badges: [RpgBadge]?
    var quests: [RpgAdminQuest]?
    var events: [RpgAdminEvent]?
    var atlas: RpgAtlas?
    var classAffinity: [String: Int]?
    var classes: [RpgClassInfo]?
    var catalogBadges: [RpgBadge]?

    enum CodingKeys: String, CodingKey {
        case player, badges, quests, events, atlas, classes
        case classAffinity = "class_affinity"
        case catalogBadges = "catalog_badges"
    }
}

struct RpgAdminQuest: Decodable, Identifiable {
    var key: String?
    var kind: String?
    var title: String?
    var status: String?
    var progress: Int?
    var target: Int?
    var rewardXp: Int?
    var periodKey: String?
    var id: String { "\(key ?? "")-\(periodKey ?? "")-\(title ?? UUID().uuidString)" }
    enum CodingKeys: String, CodingKey {
        case key, kind, title, status, progress, target
        case rewardXp = "reward_xp"
        case periodKey = "period_key"
    }
}

struct RpgAdminEvent: Decodable, Identifiable {
    var kind: String?
    var createdAt: String?
    var id: String { "\(kind ?? "")-\(createdAt ?? UUID().uuidString)" }
    enum CodingKeys: String, CodingKey {
        case kind
        case createdAt = "created_at"
    }
}

struct RpgAdminBadgeActionResponse: Decodable {
    var ok: Bool?
    var granted: Bool?
    var removed: Bool?
    var player: RpgAdminPlayerDetail?
}

/// Manifest portail `/mobile/wine/versions.json` — versions natives publiées.
struct MobileVersionsManifest: Decodable {
    var ios: String?
    var iosBuild: String?
    var android: String?
    var androidBuild: String?
    var webapp: String?
    var updatedAt: String?
    var portalUrl: String?

    enum CodingKeys: String, CodingKey {
        case ios, android, webapp
        case iosBuild = "ios_build"
        case androidBuild = "android_build"
        case updatedAt = "updated_at"
        case portalUrl = "portal_url"
    }
}

struct AdminFeedbackStats: Decodable {
    var total: Int?
    var unread: Int?
    var open: Int?
    var done: Int?
    var rejected: Int?
}

struct AdminFeedbackListResponse: Decodable {
    var items: [AdminFeedbackItem]?
    var stats: AdminFeedbackStats?
}

struct FeedbackRepliesResponse: Decodable {
    var ok: Bool?
    var items: [AdminFeedbackItem]?
    var count: Int?
}

struct AdminFeedbackItem: Decodable, Identifiable {
    var id: Int?
    var message: String?
    var category: String?
    var categoryLabel: String?
    var username: String?
    var isInvite: Bool?
    var adminRead: Bool?
    var createdAt: String?
    var clientIp: String?
    var appVersion: String?
    var device: String?
    var osName: String?
    var browser: String?
    var pagePath: String?
    var meta: [String: FlexibleJSON]?
    var status: String?
    var statusLabel: String?
    var adminReply: String?
    var resolvedAt: String?
    var resolvedBy: String?
    var userSeenReply: Bool?

    enum CodingKeys: String, CodingKey {
        case id, message, category, username, device, browser, meta, status
        case categoryLabel = "category_label"
        case isInvite = "is_invite"
        case adminRead = "admin_read"
        case createdAt = "created_at"
        case clientIp = "client_ip"
        case appVersion = "app_version"
        case osName = "os_name"
        case pagePath = "page_path"
        case statusLabel = "status_label"
        case adminReply = "admin_reply"
        case resolvedAt = "resolved_at"
        case resolvedBy = "resolved_by"
        case userSeenReply = "user_seen_reply"
    }

    /// Identifiant stable pour ForEach
    var stableId: Int { id ?? 0 }

    var displayCategory: String {
        let lab = (categoryLabel ?? "").trimmingCharacters(in: .whitespaces)
        if !lab.isEmpty { return lab }
        return category ?? "général"
    }

    var displayStatus: String {
        let lab = (statusLabel ?? "").trimmingCharacters(in: .whitespaces)
        if !lab.isEmpty { return lab }
        switch (status ?? "open").lowercased() {
        case "done": return "Mis en place"
        case "rejected": return "Refusé"
        default: return "En cours"
        }
    }

    var isOpen: Bool { (status ?? "open").lowercased() == "open" || status == nil || status?.isEmpty == true }
    var isDone: Bool { (status ?? "").lowercased() == "done" }
    var isRejected: Bool { (status ?? "").lowercased() == "rejected" }

    var deviceLine: String {
        [device, osName, browser]
            .compactMap { $0 }
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .filter { !$0.isEmpty && $0 != "—" }
            .joined(separator: " · ")
    }

    var metaRpgLevel: String? {
        guard let m = meta, let v = m["rpg_level"] else { return nil }
        switch v {
        case .int(let i): return "Lv \(i)"
        case .double(let d): return "Lv \(Int(d))"
        case .string(let s) where !s.isEmpty: return "Lv \(s)"
        default: return nil
        }
    }
}

/// JSON hétérogène pour meta feedback (int/string/bool…).
enum FlexibleJSON: Decodable {
    case string(String)
    case int(Int)
    case double(Double)
    case bool(Bool)
    case null
    case other

    init(from decoder: Decoder) throws {
        let c = try decoder.singleValueContainer()
        if c.decodeNil() { self = .null; return }
        if let i = try? c.decode(Int.self) { self = .int(i); return }
        if let d = try? c.decode(Double.self) { self = .double(d); return }
        if let b = try? c.decode(Bool.self) { self = .bool(b); return }
        if let s = try? c.decode(String.self) { self = .string(s); return }
        self = .other
    }
}

/// Compare versions "4.4.7" style. -1 si a<b, 0 égal, 1 si a>b.
func beerVersionCompare(_ a: String, _ b: String) -> Int {
    func extract(_ s: String) -> [Int] {
        // digits and dots only, first sequence
        var buf = ""
        var started = false
        for ch in s {
            if ch.isNumber {
                buf.append(ch)
                started = true
            } else if ch == "." && started {
                buf.append(ch)
            } else if started {
                break
            }
        }
        return buf.split(separator: ".").compactMap { Int($0) }
    }
    let pa = extract(a)
    let pb = extract(b)
    let n = max(pa.count, pb.count)
    for i in 0..<n {
        let x = i < pa.count ? pa[i] : 0
        let y = i < pb.count ? pb[i] : 0
        if x < y { return -1 }
        if x > y { return 1 }
    }
    return 0
}

func rarityLabelFr(_ r: String?) -> String {
    switch (r ?? "common").lowercased() {
    case "legendary": return "Légendaire"
    case "epic": return "Épique"
    case "rare": return "Rare"
    default: return "Commun"
    }
}
